consensus/event_loop.go

```diff
@@ -0,0 +1,908 @@
+		glog.Errorf("FastHotStuffEventLoop.tryConstructVoteQCInCurrentView: Failed to aggregate signatures: %v", err)
+		return nil
+	}
+
+	// Happy path. Construct the QC and return as an event to signal to the server.
+	return &FastHotStuffEvent{
+		EventType:      FastHotStuffEventTypeConstructVoteQC, // The event type
+		View:           fc.currentView,                       // The current view in which we can construct a block
+		TipBlockHash:   fc.tip.block.GetBlockHash(),          // Block hash for the tip, which we are extending from
+		TipBlockHeight: fc.tip.block.GetHeight(),             // Block height for the tip, which we are extending from
+		QC: &quorumCertificate{
+			blockHash: fc.tip.block.GetBlockHash(), // Block hash for the tip, which we are extending from
+			view:      fc.tip.block.GetView(),      // The view from the tip block. This is always fc.currentView - 1
+			aggregatedSignature: &aggregatedSignature{
+				signersList: signersList,        // The signers list who voted on the tip block
+				signature:   aggregateSignature, // Aggregated signature from votes on the tip block
+			},
+		},
+	}
+}
+
+// tryConstructTimeoutQCInCurrentView is a helper function that attempts to construct a timeout QC for the
+// previous view, so that it can be proposed in an empty block in the current view. The function internally performs
+// all view and timeout message validations to ensure that the resulting timeout QC is valid and extends from a
+// known safe block. If a timeout QC can be constructed, the function returns the safe block that it extends
+// from, the highQC, highQC views from validator timeout messages, signers list, and aggregate signature needed
+// to construct the timeout QC.
+//
+// This function must be called while holding the consensus instance's lock.
+func (fc *fastHotStuffEventLoop) tryConstructTimeoutQCInCurrentView() *FastHotStuffEvent {
+
+	// Fetch all timeouts for the previous view. All timeout messages for a view are aggregated and
+	// proposed in the next view. So if we want to propose a timeout QC in the current view, we need
+	// to aggregate timeouts from the previous one.
+	timeoutsByValidator := fc.timeoutsSeenByView[fc.currentView-1]
+
+	// Tracks the highQC from validators as we go along.
+	var validatorsHighQC QuorumCertificate
+
+	// Iterate through all timeouts for the previous view to find the highQC
+	for _, timeout := range timeoutsByValidator {
+		// Check if the high QC from the timeout messages is for a block in our safeBlocks slice. If not,
+		// then we have no knowledge of the block, or the block is not safe to extend from. This should never
+		// happen, but may be possible in the event we receive a timeout message at the same time the block
+		// becomes unsafe to extend from (ex: it's part of an stale reorg). We check for the edge case here to
+		// be 100% safe.
+		isSafeBlock, _, _, validatorLookup := fc.fetchSafeBlockInfo(timeout.GetHighQC().GetBlockHash())
+		if !isSafeBlock {
```

Comment:
It feels like you might benefit from one more message to tell server.go to fetch missing blocks when you encounter them. Something like "MissingBlockEvent" or something, which triggers server.go to request the block from a peer. But I wouldn't add this until later. Maybe just keep in mind that this could be useful for cases where a node missed some data.

Reply:
The server.go code will do this. When it receives a timeout or vote message that references a block that it hasn't seen before, it requests the block from the peer in-line. This keeps the flow for requesting blocks simple, and couples it with the vote/timeout message receipt